src/MergeContext.js

```diff
@@ -364,137 +599,167 @@ class MergeContext {
+       } catch (e) {
+           if (e.name === 'ErrorContext' && e.notFound()) {
+               this._logEx(e, this._stagingTag() + " not found");
+               return;
+           }
+           throw e;
+       }
     }
 
-    async _removeLabelsIf(labels) {
-        const currentLabels = await GH.getLabels(this.prNumber());
-        for (let label of labels) {
-            if (currentLabels.find(lbl => lbl.name === label) !== undefined)
-                await this._removeLabel(label);
-            else
-                this._log("_removeLabelsIf: skip non-existent " + label);
-        }
+    async _loadLabels() {
+        let labels = await GH.getLabels(this._prNumber());
+        assert(!this._labels);
+        this._labels = new Labels(labels, this._prNumber());
     }
 
-    async _removeLabel(label) {
-        try {
-            await GH.removeLabel(label, this.prNumber());
-        } catch (e) {
-            if (e.name === 'ErrorContext' && e.notFound()) {
-                Log.LogException(e, this._toString() + " removeLabel: " + label + " not found");
-                return;
-            }
-            throw e;
-        }
+    // Checks whether this PR is still open and still wants to be merged.
+    // This is a common part of staging and merging precondition checks.
+    async _checkActive() {
+        if (!this._prOpen())
+            throw this._exLostControl("unexpected closure");
+
+        if (this._labels.has(Config.mergedLabel()))
+            throw this._exLostControl("premature " + Config.mergedLabel());
     }
 
-    async _addLabel(label) {
-        const currentLabels = await GH.getLabels(this.prNumber());
-        if (currentLabels.find(lbl => lbl.name === label) !== undefined) {
-            this._log("addLabel: skip already existing " + label);
-            return;
-        }
+    // whether the PR should be staged (including re-staged)
+    async _checkStagingPreconditions() {
+        this._log("checking staging preconditions");
 
-        let params = Util.commonParams();
-        params.number = this.prNumber();
-        params.labels = [];
-        params.labels.push(label);
+        await this._checkActive();
 
-        await GH.addLabels(params);
+        // TODO: If multiple failures need labeling, label all of them.
+
+        if (this._wipPr())
+            throw this._exSuspend("work-in-progress");
+
+        if (!this._messageValid)
+            throw this._exLabeledFailure("invalid commit message", Config.failedDescriptionLabel());
+
+        if (!this._prMergeable())
+            throw this._exObviousFailure("GitHub will not be able to merge");
+
+        if (!this._approval.granted())
+            throw this._exSuspend("waiting for approval");
+
+        if (this._approval.grantedTimeout())
+            throw this._exSuspend("waiting for objections");
+
+        if (this._stagingBanned)
+            throw this._exSuspend("waiting for another staged PR");
+
+        if (this._stagedCommitWillFail)
```

Comment:
It feels like this check should go above the _approval.granted() check because, without a label, humans are likely to miss this poorly visible failure and may end up waiting for votes and other PRs instead of fixing the broken PR branch.

Reply:
Nice!